src/Catel.MVVM/Collections/FastObservableDictionary.cs

```diff
@@ -0,0 +1,1060 @@
+        ISerializable,
+        IDeserializationCallback,
+        INotifyCollectionChanged,
+        INotifyPropertyChanged,
+        ISuspendChangeNotificationsCollection
+    {
+        #region Constructors
+        public FastObservableDictionary()
+        {
+            _dict = new Dictionary<TKey, TValue>();
+            _dictIndexMapping = new Dictionary<TKey, int>();
+            _list = new List<TKey>();
+        }
+        public FastObservableDictionary(int capacity)
+        {
+            _dict = new Dictionary<TKey, TValue>(capacity);
+            _dictIndexMapping = new Dictionary<TKey, int>(capacity);
+            _list = new List<TKey>(capacity);
+        }
+        public FastObservableDictionary(IEnumerable<KeyValuePair<TKey, TValue>> originalDict)
+        {
+            if (originalDict is ICollection<KeyValuePair<TKey, TValue>> collection)
+            {
+                _dict = new Dictionary<TKey, TValue>(collection.Count);
+                _dictIndexMapping = new Dictionary<TKey, int>(collection.Count);
+                _list = new List<TKey>(collection.Count);
+            }
+            InsertMultipleValues(0, originalDict, false);
+        }
+        public FastObservableDictionary(IEqualityComparer<TKey> comparer)
+        {
+            _list = new List<TKey>();
+            _dictIndexMapping = new Dictionary<TKey, int>(comparer);
+            _dict = new Dictionary<TKey, TValue>(comparer);
+        }
+        public FastObservableDictionary(IDictionary<TKey, TValue> dictionary)
+        {
+            _dict = new Dictionary<TKey, TValue>(dictionary.Count);
+            _dictIndexMapping = new Dictionary<TKey, int>(dictionary.Count);
+            _list = new List<TKey>(dictionary.Count);
+            InsertMultipleValues(dictionary, false);
+        }
+        public FastObservableDictionary(int capacity, IEqualityComparer<TKey> comparer)
+        {
+            _list = new List<TKey>(capacity);
+            _dictIndexMapping = new Dictionary<TKey, int>(capacity, comparer);
+            _dict = new Dictionary<TKey, TValue>(capacity, comparer);
+        }
+        public FastObservableDictionary(IDictionary<TKey, TValue> dictionary, IEqualityComparer<TKey> comparer)
+        {
+            Argument.IsNotNull(nameof(dictionary), dictionary);
+
+            _dict = new Dictionary<TKey, TValue>(dictionary.Count, comparer);
+            _dictIndexMapping = new Dictionary<TKey, int>(dictionary.Count, comparer);
+            _list = new List<TKey>(dictionary.Count);
+            InsertMultipleValues(dictionary, false);
+        }
+
+
+        private FastObservableDictionary(Dictionary<TKey, TValue> dict, Dictionary<TKey, int> dictIndexMapping, List<TKey> list)
+        {
+            _dict = dict;
+            _list = list;
+            _dictIndexMapping = dictIndexMapping;
+        }
+        #endregion
+
+        #region Fields & Properties
+        private static readonly ILog Log = LogManager.GetCurrentClassLogger();
```

Comment:
Move to top  (before ctors) to stay consistent with other implementations we have